ql/src/java/org/apache/hadoop/hive/ql/ddl/table/storage/compact/AlterTableCompactOperation.java

```diff
@@ -69,40 +80,48 @@ public AlterTableCompactOperation(DDLOperationContext context, AlterTableCompact
       compactionRequest.setNumberOfBuckets(desc.getNumberOfBuckets());
     }
 
-    InitiatorBase initiatorBase = new InitiatorBase();
-    initiatorBase.setConf(context.getConf());
```

Comment:
Utility methods should not implicitly change the arguments especially things like configurations cause this makes them harder to use and can easily lead to bugs.

I believe we can modify the `CompactorUtil.initiateCompactionForPartition` method to not modify the input configuration by creating the `ValidTxnList` when needed (i.e., inside the `resolveValidWriteIds` method) unless there is a specific reason that we need this to be done at the very beginning. 

Assuming that it is really necessary to do the initialization of `ValidTxnList` here then we could just propagate this object as it is to subsequent methods by adding an additional param.

A final option would be to make a copy of the whole config object right before modifying it but this would be costly and wasteful in terms of resources.

Reply:
1. Changing conf at `resolveValidWriteIds` will disturb initiator flow, Initiator will set VALID_TXNS_KEY in its config object every time it runs before calling resolveValidWriteIds. And VALID_TXNS_KEY is not only used in `resolveValidWriteIds` but also some other methods like `AcidUtils.getAcidState(sd, writeIds, conf);`
Secondly, `resolveValidWriteIds` is getting executed for every partition, so it is better to keep common code into `initiateCompactionForPartition` itself to execute code only once.
2. We cannot add a new property in config object and pass it since other methods like 
`AcidUtils.getAcidState(sd, writeIds, conf);`
|
 `return getAcidState(fs, location, conf, writeIds, Ref.from(false), false);`
 |
 `public static AcidDirectory getAcidState(FileSystem fileSystem, Path candidateDirectory, Configuration conf,
      ValidWriteIdList writeIdList, Ref<Boolean> useFileIds, boolean ignoreEmptyFiles) throws IOException {`
 |		
`public static AcidDirectory getAcidState(FileSystem fileSystem, Path candidateDirectory, Configuration conf,
		      ValidWriteIdList writeIdList, Ref<Boolean> useFileIds, boolean ignoreEmptyFiles, Map<Path,
		      HdfsDirSnapshot> dirSnapshots) throws IOException {`		    
`ValidTxnList validTxnList = getValidTxnList(conf);`

   expects VALID_TXNS_KEY in the config, so adding another param will need change here as well
3. So choosing the option to clone config and set VALID_TXNS_KEY in it and pass it on to the rest of the methods. The new object is created only once per alter table compact command irrespective of the number of partitions and will not disturb initiator flow